src/hooks/useActionCenterSubscriber/useSwapActionSubscriber.tsx

```diff
@@ -0,0 +1,412 @@
+                },
+                status: ActionStatus.Complete,
+              }),
+            )
+            dispatch(
+              swapSlice.actions.upsertSwap({
+                ...swap,
+                status: SwapStatus.Success,
+              }),
+            )
+
+            const notificationTitle = selectSwapActionBySwapId(store.getState(), {
+              swapId: swap.id,
+            })?.title
+
+            toast({
+              title: notificationTitle,
+              status: 'success',
+            })
+
+            return
+          }
+        }
+
+        const notificationTitle = translate('notificationCenter.swapTitle', {
+          sellAmountAndSymbol: toCrypto(
+            fromBaseUnit(swap.sellAmountCryptoBaseUnit, swapSellAsset.precision),
+            swapSellAsset.symbol,
+            {
+              maximumFractionDigits: 8,
+              omitDecimalTrailingZeros: true,
+              abbreviated: true,
+              truncateLargeNumbers: true,
+            },
+          ),
+          buyAmountAndSymbol: toCrypto(
+            fromBaseUnit(swap.buyAmountCryptoBaseUnit, swapBuyAsset.precision),
+            swapBuyAsset.symbol,
+            {
+              maximumFractionDigits: 8,
+              omitDecimalTrailingZeros: true,
+              abbreviated: true,
+              truncateLargeNumbers: true,
+            },
+          ),
+        })
+        dispatch(
+          actionSlice.actions.upsertAction({
+            ...action,
+            swapMetadata: {
+              swapId: swap.id,
+            },
+            status: ActionStatus.Complete,
+          }),
+        )
+        dispatch(
+          swapSlice.actions.upsertSwap({
+            ...swap,
+            status: SwapStatus.Success,
+          }),
+        )
+
+        toast({
+          title: notificationTitle,
+          status: 'success',
+        })
+      }
+
+      if (status === TxStatus.Failed) {
+        dispatch(
+          actionSlice.actions.upsertAction({
+            ...action,
+            title: translate('notificationCenter.swapTitle'),
+            status: ActionStatus.Failed,
+            swapMetadata: {
+              swapId: swap.id,
+            },
+          }),
+        )
+        dispatch(
+          swapSlice.actions.upsertSwap({
+            ...swap,
+            status: SwapStatus.Failed,
+          }),
+        )
+
+        toast({
+          title: translate('notificationCenter.swapError'),
+          status: 'error',
+        })
+      }
+
+      return {
+        status,
+        message,
+        buyTxHash,
+      }
+    },
+    [toCrypto, translate, toast, dispatch],
+  )
+
+  // Update actions status when swap is confirmed or failed
+  const actionsQueries = useMemo(() => {
+    return pendingSwapActions
+      .map(action => {
+        const swapId = action.swapMetadata.swapId
+        if (!swapId) return undefined
+
+        const swap = swapsById[swapId]
+
+        if (!swap) return undefined
```

Comment:
q: under which circumstances does this happen? i.e doesn't a `swapAction` inherently have a matching swap? Or in other words, is this a paranoia check or does this actually happen?

Reply:
I removed them, it's maybe too much of paranoia, each swap notificiation should have a swapId

Also, it was mainly because at some point in this PR narrowing was worst, now as it's 100% swaps, swapId is defined which means it's safer!

Nice catch btw